### ISSUE TITLE
Unable to connect to world

### ISSUE THREAD

**Issue by lumpov** [TheRemote/Legendary-Java-Minecraft-Geyser-Floodgate#19]
iPad, latest Minecraft app in the same LAN I get error `Unable to connect to world`. Double check local IP, port (19132).

What I need additionally setting up? What I do wrong?

Docker script:

```
#!/bin/bash

container=minecraft

containersource=05jchambers/legendary-minecraft-geyser-floodgate:latest

docker stop $container

docker rm $container

docker run \
    --log-opt max-size=10m --log-opt max-file=3 \
    --name $container \
    --restart unless-stopped \
    -p 25565:25565 \
    -p 19132:19132/udp \
    -p 19132:19132 \
    -v ${PWD}/data:/minecraft \
    -d $containersource
```

docker ps:

```
a@d:~/dock/minecraft$ docker ps
CONTAINER ID   IMAGE                                                      COMMAND                  CREATED          STATUS                        PORTS                                                                                                                                         NAMES
1e824e0fdb90   05jchambers/legendary-minecraft-geyser-floodgate:latest    "/bin/bash /scripts/…"   26 minutes ago   Up 26 minutes                 0.0.0.0:19132->19132/tcp, :::19132->19132/tcp, 0.0.0.0:25565->25565/tcp, 0.0.0.0:19132->19132/udp, :::25565->25565/tcp, :::19132->19132/udp   minecraft
9
```

logs/latest.log
```
addme@devops:~/dock/minecraft/data/logs$ cat ./latest.log
[10:49:10] [ServerMain/INFO]: Environment: authHost='https://authserver.mojang.com', accountsHost='https://api.mojang.com', sessionHost='https://sessionserver.mojang.com', servicesHost='https://api.minecraftservices.com', name='PROD'
[10:49:13] [ServerMain/INFO]: Loaded 7 recipes
[10:49:14] [Server thread/INFO]: Starting minecraft server version 1.19.4
[10:49:14] [Server thread/WARN]: ****************************
[10:49:14] [Server thread/WARN]: YOU ARE RUNNING THIS SERVER AS AN ADMINISTRATIVE OR ROOT USER. THIS IS NOT ADVISED.
[10:49:14] [Server thread/WARN]: YOU ARE OPENING YOURSELF UP TO POTENTIAL RISKS WHEN DOING THIS.
[10:49:14] [Server thread/WARN]: FOR MORE INFORMATION, SEE https://madelinemiller.dev/blog/root-minecraft-server/
[10:49:14] [Server thread/WARN]: ****************************
[10:49:14] [Server thread/INFO]: Loading properties
[10:49:14] [Server thread/INFO]: This server is running Paper version git-Paper-458 (MC: 1.19.4) (Implementing API version 1.19.4-R0.1-SNAPSHOT) (Git: 5100822)
[10:49:14] [Server thread/INFO]: Server Ping Player Sample Count: 12
[10:49:14] [Server thread/INFO]: Using 4 threads for Netty based IO
[10:49:14] [Server thread/INFO]: [ChunkTaskScheduler] Chunk system is using 1 I/O threads, 1 worker threads, and gen parallelism of 1 threads
[10:49:15] [Server thread/INFO]: Default game type: SURVIVAL
[10:49:15] [Server thread/INFO]: Generating keypair
[10:49:15] [Server thread/INFO]: Starting Minecraft server on *:25565
[10:49:15] [Server thread/INFO]: Using epoll channel type
[10:49:15] [Server thread/INFO]: Paper: Using libdeflate (Linux x86_64) compression from Velocity.
[10:49:15] [Server thread/INFO]: Paper: Using OpenSSL 3.0.x (Linux x86_64) cipher from Velocity.
[10:49:16] [Server thread/INFO]: [floodgate] Loading server plugin floodgate v2.2.2-SNAPSHOT (b77-089b9a7)
[10:49:17] [Server thread/INFO]: [floodgate] Took 858ms to boot Floodgate
[10:49:17] [Server thread/INFO]: [ViaVersion] Loading server plugin ViaVersion v4.6.2-SNAPSHOT
[10:49:17] [Server thread/INFO]: [ViaVersion] ViaVersion 4.6.2-SNAPSHOT is now loaded. Registering protocol transformers and injecting...
[10:49:17] [Via-Mappingloader-0/INFO]: [ViaVersion] Loading block connection mappings ...
[10:49:18] [Server thread/INFO]: [Geyser-Spigot] Loading server plugin Geyser-Spigot v2.1.0-SNAPSHOT
[10:49:18] [Via-Mappingloader-0/INFO]: [ViaVersion] Using FastUtil Long2ObjectOpenHashMap for block connections
[10:49:19] [Server thread/INFO]: [Geyser-Spigot] Loading extensions...
[10:49:19] [Server thread/INFO]: [Geyser-Spigot] Loaded 0 extension(s)
[10:49:19] [Server thread/INFO]: Server permissions file permissions.yml is empty, ignoring it
[10:49:19] [Server thread/INFO]: Preparing level "world"
[10:49:20] [Server thread/INFO]: [floodgate] Enabling floodgate v2.2.2-SNAPSHOT (b77-089b9a7)
[10:49:20] [Server thread/INFO]: [ViaVersion] Enabling ViaVersion v4.6.2-SNAPSHOT
[10:49:20] [Server thread/INFO]: [ViaVersion] ViaVersion detected server version: 1.19.4 (762)
[10:49:21] [Server thread/WARN]: [ViaVersion] ViaVersion does not have any compatible versions for this server version!
[10:49:21] [Server thread/WARN]: [ViaVersion] Please remember that ViaVersion only adds support for versions newer than the server version.
[10:49:21] [Server thread/WARN]: [ViaVersion] If you need support for older versions you may need to use one or more ViaVersion addons too.
[10:49:21] [Server thread/WARN]: [ViaVersion] In that case please read the ViaVersion resource page carefully or use https://jo0001.github.io/ViaSetup
[10:49:21] [Server thread/WARN]: [ViaVersion] and if you're still unsure, feel free to join our Discord-Server for further assistance.
[10:49:21] [Server thread/INFO]: [Geyser-Spigot] Enabling Geyser-Spigot v2.1.0-SNAPSHOT
[10:49:21] [Server thread/INFO]: [Geyser-Spigot] ******************************************
[10:49:21] [Server thread/INFO]: [Geyser-Spigot] 
[10:49:21] [Server thread/INFO]: [Geyser-Spigot] Loading Geyser version 2.1.0-SNAPSHOT (git-master-021ffe2)
[10:49:21] [Server thread/INFO]: [Geyser-Spigot] 
[10:49:21] [Server thread/INFO]: [Geyser-Spigot] ******************************************
[10:49:29] [Network Listener - #3/INFO]: [Geyser-Spigot] Started Geyser on 0.0.0.0:19132
[10:49:29] [Server thread/INFO]: [Geyser-Spigot] Done (8.827s)! Run /geyser help for help!
[10:49:30] [Server thread/INFO]: Running delayed init tasks
[10:49:30] [Server thread/INFO]: Done (15.703s)! For help, type "help"
[10:49:30] [Server thread/INFO]: Timings Reset
[10:49:30] [Craft Scheduler Thread - 2 - ViaVersion/INFO]: [ViaVersion] Finished mapping loading, shutting down loader executor!
[10:49:30] [Server thread/WARN]: [ViaVersion] You are running a development version of the plugin, please report any bugs to GitHub.
```

**Comment by TheRemote** [comment]
If your host name is devops then I assume this is Azure and Hyper-V.  I wouldn't need to see the hostname though to identify it as either the Oracle free VM or Hyper-V because they're the only ones that do this.  

There's an entire section of the README dedicated to using Hyper-V:

```
There is a weird bug in Hyper-V that breaks UDP connections on the Minecraft server. There are two fixes for this. The simplest fix is that you have to use a Generation 1 VM with the Legacy LAN network driver.
See the following links:

    https://jamesachambers.com/minecraft-bedrock-edition-ubuntu-dedicated-server-guide/comment-page-54/#comment-13863
    https://jamesachambers.com/minecraft-bedrock-edition-ubuntu-dedicated-server-guide/comment-page-56/#comment-14207

There is a second fix that was [shared by bpsimons here](https://jamesachambers.com/legendary-minecraft-bedrock-container/comment-page-3/#comment-14654).
You need to install ethtool first with sudo apt install ethtool. Next in your /etc/network/interfaces file add "offload-tx off" to the bottom as the issue appears to be with TX offloading.
Here's an example:

# The primary network interface
auto eth0
iface eth0 inet static
address 192.168.1.5
netmask 255.255.255.0
network 192.168.1.0
broadcast 192.168.1.255
gateway 192.168.1.1
offload-tx off

This can also be done non-persistently with the following ethtool command:

ethtool -K eth0 tx off
```

Go ahead and make those changes to fix the Hyper-V buggy network stack.  These issues are exclusive to Hyper-V as well as using the free Oracle VM (meaning it's the platform that is the problem).  

That should take care of it!

**Comment by lumpov** [comment]
Thank you for your answer. I use ubuntu 18 as a host for docker images running on home lab pc.

Sorry, I'm not understand what for I need to change network interface settings to static in this example.

I try to use Minecraft server in LAN. There is no UDP or other restrictions.

**Comment by TheRemote** [comment]
Whoa.  Just as a heads up I developed the container on Ubuntu 22.04.  Ubuntu 18.04 is end of life in April but it is ancient.

What does your:

`docker --version`

command give?  Mine is 20.10.21 on my desktop:

```
james@pop-os:/media/james$ docker --version
Docker version 20.10.21, build 20.10.21-0ubuntu1~22.04.2
```

It was never intended to be run on such an old operating system.  This would be way too old to run Bedrock although it could probably handle Java.  You'd be stuck on OpenJDK 17 since the system is too old to install a newer one.

To give you an idea how old this is Bedrock stopped working on 20.04 a year ago and you are on 18.04.  It hasn't worked on 18.04 for years.  I know this isn't Bedrock but it doesn't matter.  The reason Bedrock doesn't work at all is because your OS components are so old.  It's going to be a similar thing with Docker.  

I honestly don't intend to support such an old operating system.  It's EOL on April 30th.  There's no point in fixing it for Ubuntu 18.04.  You will stop even receiving patches of any kind on April 30th 2023.  This is unsafe just like when Windows versions go end of life.   You are basically running the equivalent of Windows 8 here or something like that just with Ubuntu.  It's several versions outdated.

If we spent a whole bunch of time trying to figure out which old component on your system is causing trouble it would be completely irrelevant about a month from now when your OS will no longer receive updates ever again.  

The latest LTS version of Ubuntu is 22.04 so you are literally 2 LTS releases behind here.  I would not try running Docker containers on such a system and expect it to work.

I'm honestly surprised it was able to download the server.  Most people using OS and Docker versions this old can't even download the server.  Their DNS literally won't function.  Yours got further than that but probably because it's LTS.  The solution for all of them was simply upgrading to the current version of their operating system.

If you won't take my word for it and upgrade here's Ubuntu's word: https://ubuntu.com/blog/ubuntu-18-04-eol-for-devices.  That version has been released for 5 years and time is up.

Hopefully that helps!
